### PR TITLE
[Merged by Bors] - feat(Data/Set/Function): lemmas about inj/bij/surjectivity

### PR DESCRIPTION
**Pull request by apnelson1**
This PR adds a few API lemmas about `InjOn`, `BijOn` and `SurjOn`. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Function.lean [hunk 1]
```diff
@@ -769,6 +777,22 @@ theorem _root_.Disjoint.image {s t u : Set α} {f : α → β} (h : Disjoint s t
   rw [← hf.image_inter hs ht, h, image_empty]
 #align disjoint.image Disjoint.image
 
+lemma InjOn.image_diff {t : Set α} (h : s.InjOn f) : f '' (s \ t) = f '' s \ f '' (s ∩ t) := by
+  refine subset_antisymm (subset_diff.2 ⟨image_subset f (diff_subset s t),?_⟩)
+    (diff_subset_iff.2 (by rw [← image_union, inter_union_diff]))
+  exact Disjoint.image disjoint_sdiff_inter h (diff_subset _ _) (inter_subset_left _ _)
+
+lemma InjOn.image_diff_of_subset {f : α → β} {t : Set α } (h : InjOn f s) (hst : t ⊆ s) :
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Function.lean:785]:
```suggestion
lemma InjOn.image_diff_of_subset {f : α → β} {t : Set α} (h : InjOn f s) (hst : t ⊆ s) :
```

File: Mathlib/Data/Set/Function.lean [hunk 2]
```diff
@@ -769,6 +777,22 @@ theorem _root_.Disjoint.image {s t u : Set α} {f : α → β} (h : Disjoint s t
...
+    f '' (s \ t) = f '' s \ f '' t := by
+  rw [h.image_diff, inter_eq_self_of_subset_right hst]
+
+theorem InjOn.imageFactorization_injective (h : InjOn f s) :
+    Injective (s.imageFactorization f) :=
+  fun ⟨x,hx⟩ ⟨y,hy⟩ h' ↦ by simpa [imageFactorization, h.eq_iff hx hy] using h'
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Function.lean:791]:
```suggestion
  fun ⟨x, hx⟩ ⟨y, hy⟩ h' ↦ by simpa [imageFactorization, h.eq_iff hx hy] using h'
```

File: Mathlib/Data/Set/Function.lean [hunk 3]
```diff
@@ -1360,6 +1390,14 @@ theorem SurjOn.mapsTo_invFunOn [Nonempty α] (h : SurjOn f s t) : MapsTo (invFun
   fun _y hy => mem_preimage.2 <| invFunOn_mem <| h hy
 #align set.surj_on.maps_to_inv_fun_on Set.SurjOn.mapsTo_invFunOn
 
+theorem SurjOn.image_invFun_image_subset [Nonempty α] {r : Set β} (hf : SurjOn f s t)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Function.lean:1393]: These two theorems should have invFunOn in the name rather than invFun, I think

File: Mathlib/Data/Set/Function.lean [hunk 4]
```diff
@@ -769,6 +777,22 @@ theorem _root_.Disjoint.image {s t u : Set α} {f : α → β} (h : Disjoint s t
   rw [← hf.image_inter hs ht, h, image_empty]
 #align disjoint.image Disjoint.image
 
+lemma InjOn.image_diff {t : Set α} (h : s.InjOn f) : f '' (s \ t) = f '' s \ f '' (s ∩ t) := by
+  refine subset_antisymm (subset_diff.2 ⟨image_subset f (diff_subset s t),?_⟩)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Function.lean:781]:
```suggestion
  refine subset_antisymm (subset_diff.2 ⟨image_subset f (diff_subset s t), ?_⟩)
```

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/Data/Set/Function.lean [hunk 5]
```diff
@@ -760,7 +760,15 @@ lemma InjOn.image_inter {s t u : Set α} (hf : u.InjOn f) (hs : s ⊆ u) (ht : t
   exact ⟨y, ⟨ys, zt⟩, hy⟩
 #align set.inj_on.image_inter Set.InjOn.image_inter
 
-end injOn
+lemma InjOn.image_eq_image_iff_of_subset (h : s.InjOn f) (h₁ : s₁ ⊆ s) (h₂ : s₂ ⊆ s) :
+    f '' s₁ = f '' s₂ ↔ s₁ = s₂ :=
+  ⟨fun h' ↦ by rw [← h.preimage_image_inter h₁, h', h.preimage_image_inter h₂], fun h' ↦ by rw [h']⟩
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:765]:
```suggestion
lemma InjOn.image (h : s.InjOn f) (h₁ : s₁ ⊆ s) (h₂ : s₂ ⊆ s) :
    s.Powerset..InjOn (image f) :=
  fun s₁ hs₁ s₂ hs₂ h' ↦ by rw [← h.preimage_image_inter h₁, h', h.preimage_image_inter h₂]
```

Comment (apnelson1) [reply Mathlib/Data/Set/Function.lean:765]: I think this can wait for a larger `image`/`Powerset` PR.

**Comment by YaelDillies** [comment]
Don't add the `maintainer-merge` label by hand! That's the safest way for your PR to go through the cracks.

File: Mathlib/Data/Set/Function.lean [hunk 6]
```diff
@@ -769,6 +780,22 @@ theorem _root_.Disjoint.image {s t u : Set α} {f : α → β} (h : Disjoint s t
...
+lemma InjOn.image_diff_of_subset {f : α → β} {t : Set α} (h : InjOn f s) (hst : t ⊆ s) :
+    f '' (s \ t) = f '' s \ f '' t := by
+  rw [h.image_diff, inter_eq_self_of_subset_right hst]
+
+theorem InjOn.imageFactorization_injective (h : InjOn f s) :
+    Injective (s.imageFactorization f) :=
+  fun ⟨x, hx⟩ ⟨y, hy⟩ h' ↦ by simpa [imageFactorization, h.eq_iff hx hy] using h'
+
+end injOn
+
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:797]:
```suggestion
```

File: Mathlib/Data/Set/Function.lean [hunk 7]
```diff
@@ -769,6 +780,22 @@ theorem _root_.Disjoint.image {s t u : Set α} {f : α → β} (h : Disjoint s t
...
+  refine subset_antisymm (subset_diff.2 ⟨image_subset f (diff_subset s t), ?_⟩)
+    (diff_subset_iff.2 (by rw [← image_union, inter_union_diff]))
+  exact Disjoint.image disjoint_sdiff_inter h (diff_subset _ _) (inter_subset_left _ _)
+
+lemma InjOn.image_diff_of_subset {f : α → β} {t : Set α} (h : InjOn f s) (hst : t ⊆ s) :
+    f '' (s \ t) = f '' s \ f '' t := by
+  rw [h.image_diff, inter_eq_self_of_subset_right hst]
+
+theorem InjOn.imageFactorization_injective (h : InjOn f s) :
+    Injective (s.imageFactorization f) :=
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:793]: Is this an iff?

File: Mathlib/Data/Set/Function.lean [hunk 8]
```diff
@@ -1360,6 +1393,14 @@ theorem SurjOn.mapsTo_invFunOn [Nonempty α] (h : SurjOn f s t) : MapsTo (invFun
   fun _y hy => mem_preimage.2 <| invFunOn_mem <| h hy
 #align set.surj_on.maps_to_inv_fun_on Set.SurjOn.mapsTo_invFunOn
 
+theorem SurjOn.image_invFunOn_image_subset [Nonempty α] {r : Set β} (hf : SurjOn f s t)
+    (hrt : r ⊆ t) : f '' ((Function.invFunOn f s) '' r) = r :=
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:1397]:
```suggestion
    (hrt : r ⊆ t) : f '' f.invFunOn s '' r = r :=
```

File: Mathlib/Data/Set/Function.lean [hunk 9]
```diff
@@ -1360,6 +1393,14 @@ theorem SurjOn.mapsTo_invFunOn [Nonempty α] (h : SurjOn f s t) : MapsTo (invFun
...
+  hf.rightInvOn_invFunOn.image_image' hrt
+
+theorem SurjOn.image_invFunOn_image [Nonempty α] (hf : SurjOn f s t) :
+    f '' ((Function.invFunOn f s) '' t) = t :=
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:1401]:
```suggestion
    f '' f.invFunOn s '' t = t :=
```

File: Mathlib/Data/Set/Function.lean [hunk 10]
```diff
@@ -769,6 +780,22 @@ theorem _root_.Disjoint.image {s t u : Set α} {f : α → β} (h : Disjoint s t
   rw [← hf.image_inter hs ht, h, image_empty]
 #align disjoint.image Disjoint.image
 
+lemma InjOn.image_diff {t : Set α} (h : s.InjOn f) : f '' (s \ t) = f '' s \ f '' (s ∩ t) := by
+  refine subset_antisymm (subset_diff.2 ⟨image_subset f (diff_subset s t), ?_⟩)
+    (diff_subset_iff.2 (by rw [← image_union, inter_union_diff]))
+  exact Disjoint.image disjoint_sdiff_inter h (diff_subset _ _) (inter_subset_left _ _)
+
+lemma InjOn.image_diff_of_subset {f : α → β} {t : Set α} (h : InjOn f s) (hst : t ⊆ s) :
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:788]: This is the lemma I would call `image_diff`
```suggestion
lemma InjOn.image_diff {f : α → β} {t : Set α} (h : InjOn f s) (hst : t ⊆ s) :
```

File: Mathlib/Data/Set/Function.lean [hunk 11]
```diff
@@ -769,6 +780,22 @@ theorem _root_.Disjoint.image {s t u : Set α} {f : α → β} (h : Disjoint s t
   rw [← hf.image_inter hs ht, h, image_empty]
 #align disjoint.image Disjoint.image
 
+lemma InjOn.image_diff {t : Set α} (h : s.InjOn f) : f '' (s \ t) = f '' s \ f '' (s ∩ t) := by
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:783]: and this is the lemma I would consider "auxiliary"
```suggestion
lemma InjOn.image_diff' {t : Set α} (h : s.InjOn f) : f '' (s \ t) = f '' s \ f '' (s ∩ t) := by
```

Comment (apnelson1) [reply Mathlib/Data/Set/Function.lean:783]: I disagree.

File: Mathlib/Data/Set/Function.lean (hunk 7 above)
Comment (apnelson1) [reply Mathlib/Data/Set/Function.lean:793]: Maybe, but applying this in the other direction feels contrived to me.

**Comment by apnelson1** [comment]
> Don't add the `maintainer-merge` label by hand! That's the safest way for your PR to go through the cracks.

I don't understand the control flow here, but that label was added earlier by @Ruben-VandeVelde . I believe I have responded to comments, and would like to see this merged.

**Comment by YaelDillies** [comment]
The control flow is that, when someone `maintainer merge`s a PR, the bot sends a message to a private stream. Then either a maintainer merges the PR or a reviewer points something out. In either case, they react to the bot's message with an emoji to indicate that no more attention is needed.

By changing the label by hand, you're taking off the PR from the queue _without_ the bot sending a message to the private stream. In short, you're making your PR invisible.

File: Mathlib/Data/Set/Function.lean (hunk 7 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean:793]: Please avoid creating consensus by closing review comments you didn't address. Maintainers can always decide whether those comments are worth discarding.

File: Mathlib/Data/Set/Function.lean (hunk 8 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean]: Why are you keeping the brackets?

File: Mathlib/Data/Set/Function.lean (hunk 9 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean]: Same here

File: Mathlib/Data/Set/Function.lean (hunk 10 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean:788]: What about `image_diff_subset`?

File: Mathlib/Data/Set/Function.lean [hunk 12]
```diff
@@ -1360,6 +1392,14 @@ theorem SurjOn.mapsTo_invFunOn [Nonempty α] (h : SurjOn f s t) : MapsTo (invFun
...
 #align set.surj_on.maps_to_inv_fun_on Set.SurjOn.mapsTo_invFunOn
 
+theorem SurjOn.image_invFunOn_image_subset [Nonempty α] {r : Set β} (hf : SurjOn f s t)
+    (hrt : r ⊆ t) : f '' ((f.invFunOn s) '' r) = r :=
+  hf.rightInvOn_invFunOn.image_image' hrt
+
+theorem SurjOn.image_invFunOn_image [Nonempty α] (hf : SurjOn f s t) :
+    f '' ((f.invFunOn s) '' t) = t :=
+  hf.rightInvOn_invFunOn.image_image
+
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:1402]: I was going to say that those lemmas should be replaced by `RightInvOn` statements as this is what the API expects you to use and I was going to suggest you the proofs. But I see you already found the proofs so you might as well drop the lemmas. We don't want one lemma of the form `f '' g '' s = s` for every possible pair of functions `f` and `g` that are inverse to each other on some set `s`. These lemmas are easily accessed on the fly using the API.

File: Mathlib/Data/Set/Function.lean (hunk 7 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean:793]: > Maybe, but applying this in the other direction feels contrived to me. 

Iffs do not necessarily need to be applied the other way to be useful. A great use of them is rewriting.

File: Mathlib/Data/Set/Function.lean (hunk 12 above)
Comment (apnelson1) [reply Mathlib/Data/Set/Function.lean:1402]: Easily accessed only once you know they're there. I think this special case is prominent enough for these lemmas to exist so they're easily findable.

File: Mathlib/Data/Set/Function.lean (hunk 7 above)
Comment (apnelson1) [reply Mathlib/Data/Set/Function.lean:793]: You lose dot notation this way. Adding a separate `iff` lemma is possible, but it feels almost useless to me.

File: Mathlib/Data/Set/Function.lean (hunk 12 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean:1402]: Will let someone else weigh in

File: Mathlib/Data/Set/Function.lean (hunk 7 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean:793]: Two lemmas sounds like the way to go to me

File: Mathlib/Data/Set/Function.lean (hunk 12 above)
Comment (apnelson1) [reply Mathlib/Data/Set/Function.lean:1402]: Of course the lemmas are already there at the right generality, for me it is about the analogy between `InjOn` and `SurjOn`. It feels counterintuitive that for one, certain lemmas are there, but for others the lemmas that look like they should be there actually aren't. I think that it's best not to require that people figure out the reason for this. 

I'm also not sure how far through the API `exact?` can see in this case, but possibly this helps with that as well.

**Comment by madvorak** [comment]
I find it a bit hard to parse `f '' (s \ t) = f '' s \ f '' (s ∩ t)` optically.
I know that redundant parentheses are unpopular, but I personally would like to see 
`f '' (s \ t) = (f '' s) \ (f '' (s ∩ t))` here.

**Review by madvorak — APPROVED** [review]

**Comment by YaelDillies** [comment]
maintainer merge

but note you will have to please the style linter by running `./scripts/update_style_exceptions.sh`

File: Mathlib/Data/Set/Function.lean [hunk 13]
```diff
@@ -769,6 +785,25 @@ theorem _root_.Disjoint.image {s t u : Set α} {f : α → β} (h : Disjoint s t
   rw [← hf.image_inter hs ht, h, image_empty]
 #align disjoint.image Disjoint.image
 
+lemma InjOn.image_diff {t : Set α} (h : s.InjOn f) : f '' (s \ t) = f '' s \ f '' (s ∩ t) := by
+  refine subset_antisymm (subset_diff.2 ⟨image_subset f diff_subset, ?_⟩)
+    (diff_subset_iff.2 (by rw [← image_union, inter_union_diff]))
+  exact Disjoint.image disjoint_sdiff_inter h diff_subset inter_subset_left
+
+lemma InjOn.image_diff_subset {f : α → β} {t : Set α} (h : InjOn f s) (hst : t ⊆ s) :
```
Comment (fpvandoorn) [inline Mathlib/Data/Set/Function.lean:793]:
```suggestion
lemma InjOn.image_diff_of_subset {f : α → β} {t : Set α} (h : InjOn f s) (hst : t ⊆ s) :
```

File: Mathlib/Data/Set/Function.lean [hunk 14]
```diff
@@ -760,7 +760,23 @@ lemma InjOn.image_inter {s t u : Set α} (hf : u.InjOn f) (hs : s ⊆ u) (ht : t
   exact ⟨y, ⟨ys, zt⟩, hy⟩
 #align set.inj_on.image_inter Set.InjOn.image_inter
 
-end injOn
+lemma InjOn.image (h : s.InjOn f) : s.powerset.InjOn (image f) :=
+  fun s₁ hs₁ s₂ hs₂ h' ↦ by rw [← h.preimage_image_inter hs₁, h', h.preimage_image_inter hs₂]
+
+theorem InjOn.image_eq_image_iff_of_subset (h : s.InjOn f) (h₁ : s₁ ⊆ s) (h₂ : s₂ ⊆ s) :
```
Comment (fpvandoorn) [inline Mathlib/Data/Set/Function.lean:766]: for these and the next two lemmas, it's probably unambiguous to remove `_of_subset`?

**Review by fpvandoorn — APPROVED** [review]
Few comments, otherwise looks good to me.

bors d+

File: Mathlib/Data/Set/Function.lean (hunk 14 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean:766]: possibly even use `image_inj` instead of `image_eq_image`

File: Mathlib/Data/Set/Function.lean (hunk 13 above)
Comment (fpvandoorn) [reply Mathlib/Data/Set/Function.lean:792]: @apnelson1 I think you missed this comment?
